### PR TITLE
feat(backup): support mutating backup config

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR makes system params `backup_storage_url` and `backup_storage_directory` mutable.
These 2 params are used by both meta node and compute node. They are updated in a similar way:
- watch system params changes.
- recreate underlying storage that stores meta snapshot, if any of the 2 params has been changed.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/storage/src/hummock/backup_reader.rs [hunk 1]
```diff
@@ -184,6 +214,31 @@ impl BackupReader {
...
+    ) {
+        loop {
+            if rx.changed().await.is_err() {
+                break;
+            }
+            let p = (*rx.borrow()).load();
+            let config = (
+                p.backup_storage_url().to_string(),
+                p.backup_storage_directory().to_string(),
+            );
```
Comment (Gun9niR) [inline src/storage/src/hummock/backup_reader.rs:230]: `rx.changed().await` will return when **any** system param changes, so we need to check if these two params differ from the old ones. 

Or do you think we should have a separate channel for each parameter, to make the interface more friendly?

File: src/storage/src/hummock/backup_reader.rs [hunk 2]
```diff
@@ -184,6 +214,34 @@ impl BackupReader {
...
+
+    pub async fn watch_config_change(
+        &self,
+        mut rx: tokio::sync::watch::Receiver<SystemParamsReaderRef>,
+    ) {
+        loop {
+            if rx.changed().await.is_err() {
+                break;
+            }
+            let p = (*rx.borrow()).load();
```
Comment (Gun9niR) [inline src/storage/src/hummock/backup_reader.rs:226]:
```suggestion
            let p = rx.borrow().load();
```

**Review by Gun9niR — APPROVED** [review]
LGTM

**Comment by zwang28** [comment]
> Will there be any problem if different CNs receive this update at different time?

No.

This new config only affects ad-hoc query. If two CNs use old and new backup store respectively, it falls into either of the 2 cases:
- The target backup identified by backup id can be found in both stores. **We assume two backups are the same, if their backup id are same.**
- The target backup can be found in only one store. The corresponding ad-hoc query will fail as expected.

**Pull request merged**